### DISCUSSION TITLE
Is Java grammar somehow finished?

### DISCUSSION THREAD

**Discussion by dims12** [antlr/grammars-v4#2897] [Q&A]
Hi! I noticed that `additiveExpression` is never used here: https://github.com/antlr/grammars-v4/blob/master/java/java8/Java8Parser.g4#L1276`

(Used only in itself and in `shiftExpression`)

Is it by design of Java grammar is far from completeness?

**Comment by msagca** [comment]
Hi @dims12 

`additiveExpression` is referenced by `shiftExpression` which is referenced by `relationalExpression` which is referenced by `equalityExpression` and so on. The grammar is designed this way because the operator precedence is important in order for the resulting parse tree to be correct. `shiftExpression` can be unfolded as follows:

```antlr
shiftExpression
: unaryExpression
| shiftExpression ('*' | '/' | '%') shiftExpression
| shiftExpression ('+' | '-') shiftExpression
| shiftExpression ('<<' | '>>' | '>>>') shiftExpression
;
```

Similarly, we can rewrite the `expression` rule by performing unfold transforms and get the following more compact form:

```antlr
expression
: lambdaExpression
| unaryExpression
| expression ('*' | '/' | '%') expression
| expression ('+' | '-') expression
| expression ('<<' | '>>' | '>>>') expression
| expression ('<' | '>' | '<=' | '>=' | 'instanceof') expression
| expression ('==' | '!=') expression
| expression '&' expression
| expression '^' expression
| expression '|' expression
| expression '&&' expression
| expression '||' expression
| <assoc=right> expression '?' expression ':' expression
| leftHandSide ('=' | '*=' | '/=' | '%=' | '+=' | '-=' | '<<=' | '>>=' | '>>>=' | '&=' | '^=' | '|=') expression
;
```

**Comment by KvanTTT** [comment]
I don't recommend using this grammar because it's slow and outdated. Use [java](https://github.com/antlr/grammars-v4/tree/master/java/java) instead.
